sw/device/lib/base/csr.h

```diff
@@ -0,0 +1,287 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_CSR_H_
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/stdasm.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * @file
+ * @brief Control and Status Register (CSR) interface.
+ *
+ * A set of macros that provide both read and modify operations on CSRs.
+ * Compiling translation units that include this header file with `-DMOCK_CSR`
+ * will result in the CSR operations being replaced with a mocked
+ * implementation.
+ */
+
+/**
+ * Define the implementation macros.
+ *
+ * The implementation used depends on whether the CSR library is providing a
+ * real or a mocked interface.
+ */
+#ifdef MOCK_CSR
+
+uint32_t mock_csr_read(uint32_t addr);
+
+#define CSR_READ_IMPL(csr, dst)  \
+  do {                           \
+    asm volatile("" ::"i"(csr)); \
+    *dst = mock_csr_read(csr);   \
+  } while (0);
+
+void mock_csr_write(uint32_t addr, uint32_t value);
+
+#define CSR_WRITE_IMPL(csr, val) \
+  do {                           \
+    asm volatile("" ::"i"(csr)); \
+    mock_csr_write(csr, val);    \
+  } while (0);
+
+void mock_csr_set_bits(uint32_t addr, uint32_t mask);
```

Comment:
Given that this behaves like a RMW operation, perhaps it's worth implementing it in terms of read and write, to simplify downstream expectations?

Reply:
I'm not sure what you mean here about downstream expectations? The clear and set bits functions are useful in part because they are atomic.

Reply:
Sine the actual implementations use `csrs` and `crsc`, I think defining matching `EXPECT_CSR_*` macros as in this PR is good (to be able to differentiate between `CSR_SET_BITS` and `CSR_READ` + `CSR_WRITE`). I had an earlier comment regarding dedicated `SET/CLEAR_BITS` macros but it looks like we need them.